### DISCUSSION TITLE
zx doesn't seem to work as part of a node bin script

### DISCUSSION THREAD

**Discussion by bkjohnson** [google/zx#802] [Help needed] [answered]
I'm trying to have a zx script that feels like a more traditional cli tool, meaning that it is installed to the PATH so that a user can run it by just typing `my-cli command --opts` instead of `zx path/to/my-script.mjs`. To do this I'm using the [`package.json` `bin` property](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#bin). Here's the gist of my package.json:

```json
{
  "name": "my-cli",
  "version": "0.0.1",
  "bin": {
    "my-cli": "./index.js"
  },
  "dependencies": {
    "zx": "^8.0.2"
  }
}
```

And my index file is doing something very basic:

```js
#!/usr/bin/env zx

void (async function main() {
  const branch = await $`git branch --show-current`;
  echo(branch);
})();
```

From the package directory if I install this globally with `npm install -g .` I end up with the script in `/usr/bin` which is what I expect. However, I get a permissions error when I try to run it because zx is trying to open a `.mjs` file for writing.

```
[Error: EACCES: permission denied, open '/usr/bin/my-cli.mjs'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/usr/bin/my-cli.mjs'
}
```

I think I've found out where it's coming from:

https://github.com/google/zx/blob/0eb81c93d9b5637577f143ed318449b6ea9373c7/src/cli.ts#L160-L169

`tmpFilename` is `'/usr/bin/my-cli.mjs'`, and then we try to write the contents of the original script to it:

https://github.com/google/zx/blob/0eb81c93d9b5637577f143ed318449b6ea9373c7/src/cli.ts#L144-L155

One of my questions is why is it necessary to write to a new `.mjs` file? Am I trying to use zx in too unexpected of a way? This kind of setup would work without issue for a regular node script, so I wouldn't think that this would be such an edge case.

**Comment by antongolub** [comment, marked as answer]
You can avoid bin composition inside your script. As a workaround for now:
```js
#!/usr/bin/env node
import 'zx/globals'

void (async function main() {
  const branch = await $`git branch --show-current`;
  echo(branch);
})();
```

**Reply by bkjohnson** [reply]
Thank you, that worked! Is there any downside or "gotcha" to this approach?
